### DISCUSSION TITLE
User provided job system implementation issues.

### DISCUSSION THREAD

**Discussion by BobbyAnguelov** [jrouwe/JoltPhysics#457] [General]
Hi, 

So I'm integrating Jolt and have been having a fair amount of issues with providing an implementation of the job system required. I'm using enkiTS as the basis for my engine's job system and it provides pretty much everything needed on a conceptual level to provide an implementation.

The issues arise from the fact that the base "JPH::JobSystem" makes some assumptions about how a job system should work. For example, the dependencies not being explicit but rather just a count, which forces the client to basically mimic the implementation in JobSystemThreadPool (the iterating over a list and executing whatever tasks are able to run) except now forwarding the actual execution to the engine's job system. This basically results in double book-keeping on my end, where I need to build a rough scheduler to feed another scheduler. Looking through the code, I'm not really sure that given the assumptions and intentions in the architecture of jobs and the job system, that there really is any other implementation of this aside from what's already provided in the thread pool version in terms of dependency tracking and actually executing the jobs.

I also can't use the JobSystemThreadPool as is since I need to init/shutdown my memory allocator on every thread it is called from. So I think for now I'm going to hack this by copying the threadpool implementation and just initializing my allocator as required. That said it's really not ideal to have another thread pool instead of being able to just use the existing one.

Most other middleware just exposes the part of the API that is responsible for dispatching and executing a task in parallel (See PhysX task-dispatcher as an example). Perhaps the job system could be modified in future versions to handle all the dependencies tracking/barrier/queue management on jolt's side and just expose an API for executing the actual jobs as I'm not really sure that anyone will provide any implementation that differs from the threadpool one in any meaningful way.

I'm also open to any ideas on how to integrate this in a not horrible way into an existing job system. Right now every idea I have is various degrees of awful.

**Comment by jrouwe** [comment]
Hello,

The idea behind the JobSystem interface is that if you have a job scheduler that allows you to run a job without dependencies, it should work with Jolt. The dependency tracking/counting system is something that Jolt uses internally and should not have to be repeated in your own JobSystem implementation. 

You should implement JobSystem::CreateJob to create a job and JobSystem::QueueJob(s) to immediately schedule the job without any dependencies. You can copy the Barrier implementation from JobSystemThreadPool into your own JobSystem implementation, it doesn't depend on any specific job scheduler implementation (it will only execute jobs that have been added to the barrier, but you don't want any other jobs to run anyway as any interaction with the PhysicsSystem from the thread that's waiting for the barrier is likely to deadlock).

Jolt builds the job graph while the job graph is already executing (for performance reasons), so there's no way you can first create a complete list of jobs with dependencies and then start the job graph.

**Reply by BobbyAnguelov** [reply]
Okay, I guess I misunderstood the intended usage while looking at the thread pool implmentation. Will try to do the simplest thing and see if it works!

**Comment by BobbyAnguelov** [comment]
So I think I have something working but there's several things that are making me a bit uneasy with the whole thing, so I have some dumb questions:

- Is it guaranteed that all jobs created by "CreateJob" will be added to the barrier?
- Is it safe to assume there will only ever be one barrier per "PhysicsSystem"
- Is this correct: "add ref" when queuing a job, "add ref" when adding job to barrier", "release" immediately after finishing executing a job and "release" after detecting a job is finished while waiting in the barrier?

**Reply by jrouwe** [reply]
> * Is it guaranteed that all jobs created by "CreateJob" will be added to the barrier?

That should be the case, if not we potentially waste CPU cycles on the thread that's waiting for the Barrier (as it doesn't know about the job so can't run it).

> * Is it safe to assume there will only ever be one barrier per "PhysicsSystem"

That is also the case at the moment and unlikely to ever change.

> * Is this correct: "add ref" when queuing a job, "add ref" when adding job to barrier", "release" immediately after finishing executing a job and "release" after detecting a job is finished while waiting in the barrier?

The job pointer that gets passed in to all functions is valid only for the duration of that function call. If you want to store that pointer in your own data structure you should call AddRef, and when you're done with it you call Release. So, yes, your statement is correct with the caveat that after a call to Release, the memory can be freed (if you happen to hold the last reference) so be sure not to dereference the pointer anymore.

**Comment by BobbyAnguelov** [comment]
Perfect! Thanks! 

It might be a good idea to perhaps add some extra documentation (with some of these points) to the JobSystem header as it might save some people some headache in the future. I ended up basically printf debugging the job queueing to try to understand the base behavior  without getting confused by the threadpool impl.

**Reply by jrouwe** [reply]
Yes, good idea!

**Reply by jrouwe** [reply]
Done: https://github.com/jrouwe/JoltPhysics/commit/959e47dda72eb3fa2d56c55a4cefd38e81b9a2a5

Do you think you would have been able to implement it with this new documentation?

**Reply by BobbyAnguelov** [reply]
Yes, that's great thanks - that would have been super useful! 

Once I find a bit of time this week, I'll add a couple of extra comments that might also prove useful for other and make a PR :)

**Reply by jrouwe** [reply]
I split off the Barrier implementation from the JobSystemThreadPool class into a class called JobSystemWithBarrier (see #464), this should make it a lot easier for future people to implement their own job system.

**Reply by BobbyAnguelov** [reply]
Oh that will definitely make it easier :)

Thanks a lot!!!

**Reply by BobbyAnguelov** [reply]
Just an update on this, I switched over to the "JobSystemWithBarrier" and just wanted to say, this is great! Thanks again for this!

**Reply by jrouwe** [reply]
Thank you for helping me to get the docs in order!

**Reply by mihe** [reply]
I want to also say thank you for adding `JobSystemWithBarrier`. I made an attempt at implementing `JobSystem` a few months back and got overwhelmed when trying to figure out all the barrier stuff and decided to leave it for later. Now with `JobSystemWithBarrier` it was a breeze to get things working.

Thanks to the both of you!
